### PR TITLE
ERC-721 Vyper Annotated Code

### PR DESCRIPTION
**Pull request by qbzzt**
## Description

The Vyper ERC-721 contract with commentary.

## Related Issue

#1785 requests this kind of article. This is the second one (the first one was ERC-20 and Solidity)

### CODE REVIEW

**Comment by wackerow** [comment]
@qbzzt nice! Sorry for delays on this, trying to play some catch up as we're doing some planning. Will review soon but wanted to let you know it's not being ignored

**Comment by qbzzt** [comment]
> @qbzzt nice! Sorry for delays on this, trying to play some catch up as we're doing some planning. Will review soon but wanted to let you know it's not being ignored

Thank you. I figured with the Berlin hardfork coming up everybody's busy.

File: src/content/developers/tutorials/erc721-vyper-annotated-code/index.md [hunk 1]
```diff
@@ -0,0 +1,731 @@
+---
+title: "Vyper ERC-721 Contract Walk-Through"
```
Comment (wackerow) [inline src/content/developers/tutorials/erc721-vyper-annotated-code/index.md:2]: `Walkthrough`

File: src/content/developers/tutorials/erc721-vyper-annotated-code/index.md [hunk 2]
```diff
@@ -0,0 +1,731 @@
...
+
+```python
+# @dev Mapping from NFT ID to the address that owns it.
+idToOwner: HashMap[uint256, address]
+
+# @dev Mapping from NFT ID to approved address.
+idToApprovals: HashMap[uint256, address]
+```
+
+User and contract identities in Ethereum are represented by 160 bit addresses. These two variables map
```
Comment (wackerow) [inline src/content/developers/tutorials/erc721-vyper-annotated-code/index.md:182]: hyphenate `160-bit`

File: src/content/developers/tutorials/erc721-vyper-annotated-code/index.md [hunk 3]
```diff
@@ -0,0 +1,731 @@
...
+
+
+```python
+# @dev Mapping from owner address to count of his tokens.
+ownerToNFTokenCount: HashMap[address, uint256]
+```
+
+This variable holds the count of tokens for each owner. There is no mapping from owners to tokens, so
+the only way to identify the tokens that a specific owner owns is to look back in the blockchain's event history 
+and see the appropriate `Transfer`  events. We can use this variable to know when we have all the NFTs and don't
```
Comment (wackerow) [inline src/content/developers/tutorials/erc721-vyper-annotated-code/index.md:195]: extra space after `Transfer`

File: src/content/developers/tutorials/erc721-vyper-annotated-code/index.md [hunk 4]
```diff
@@ -0,0 +1,731 @@
...
+* `@view` specifies that this function is a view.
+* `@external` specifies that this particular function can be called by transactions and by other contracts.
+
+```python
+def supportsInterface(_interfaceID: bytes32) -> bool:
+```
+
+In contrast to Python, Vyper is a [static typed language](https://en.wikipedia.org/wiki/Type_system#Static_type_checking). 
+You can't declare a variable, or a function parameter, without identifying the [data 
+type](https://vyper.readthedocs.io/en/latest/types.html). In this case the input parameter is `bytes32`, a 256 bit value
```
Comment (wackerow) [inline src/content/developers/tutorials/erc721-vyper-annotated-code/index.md:292]: hyphenate `256-bit`, but keep `256 bits` in the next line... the subtleties =)

File: src/content/developers/tutorials/erc721-vyper-annotated-code/index.md [hunk 5]
```diff
@@ -0,0 +1,731 @@
...
+    """
+    # Throws if `_owner` is not the current owner
+    assert self.idToOwner[_tokenId] == _owner
+    if self.idToApprovals[_tokenId] != ZERO_ADDRESS:
+        # Reset approvals
+        self.idToApprovals[_tokenId] = ZERO_ADDRESS
+```
+
+Only change the value if necessary. State variables live in storage. Writing to storage is
+one of the most expensive operations the evm (Ethereum Virtual Machine) does (in terms of
```
Comment (wackerow) [inline src/content/developers/tutorials/erc721-vyper-annotated-code/index.md:468]: `EVM`

File: src/content/developers/tutorials/erc721-vyper-annotated-code/index.md [hunk 6]
```diff
@@ -0,0 +1,731 @@
...
+In contrast to Solidty, Vyper does not have inheritence. This is a deliberate design choice to make the 
+code clearer and therefore easier to secure. So to create your own Vyper ERC-721 contract you take [this
+contract](https://github.com/vyperlang/vyper/blob/master/examples/tokens/ERC721.vy) and modify it 
+to implement the business logic you want.
+
+
+# Conclusion {#conclusion}
+
+
+For review, here are some of the most important ideas in this contract (in my opinion, yours is likely to vary):
```
Comment (wackerow) [inline src/content/developers/tutorials/erc721-vyper-annotated-code/index.md:719]: `(in my opinion, yours is likely to vary)`
hah, don't be afraid to own it with confidence =) it's your post

File: src/content/developers/tutorials/erc721-vyper-annotated-code/index.md [hunk 7]
```diff
@@ -0,0 +1,731 @@
+---
```
Comment (wackerow) [inline src/content/developers/tutorials/erc721-vyper-annotated-code/index.md:1]: Mind just changing the filename to hyphenate erc-721?
`/erc-721-vyper-annotated-code/index.md`

**Review by wackerow — CHANGES_REQUESTED** [review]
@qbzzt Hey! Thanks for being so patient on this one... This looks great, just dropped a few semantic nit-picky adjustments and then we'll get this up.. thanks again!

**Comment by qbzzt** [comment]
> @qbzzt Hey! Thanks for being so patient on this one... This looks great, just dropped a few semantic nit-picky adjustments and then we'll get this up.. thanks again!

No problem. The changes have been done and committed, hopefully we're ready for publication

**Review by wackerow — APPROVED** [review]
@qbzzt 💪 Nice. Thanks for the touch-ups! Bringing it in 🎉

**Pull request merged**